python/whylogs/api/logger/experimental/logger/actor/process_rolling_logger_messages.py

```diff
@@ -0,0 +1,283 @@
+
+    def to_pubsub_embedding_message(self) -> Optional[PubSubEmbeddingDict]:
+        data: PubSubEmbeddingDict = orjson.loads(self.request)
+        log_request_dict_bytes = _dencode_pubsub_data(data)
+
+        if log_request_dict_bytes is None:
+            return None
+
+        log_message = RawLogEmbeddingsMessage(request=log_request_dict_bytes, request_time=self.request_time)
+        log_request = log_message.to_log_embeddings_request_dict()
+        if log_request is None:
+            return None
+        data["log_embedding_request"] = log_request
+        return data
+
+
+@dataclass
+class RawLogEmbeddingsMessage:
+    request: bytes
+    request_time: int
+
+    def to_log_embeddings_request_dict(self) -> Optional[LogEmbeddingRequestDict]:
+        d: LogEmbeddingRequestDict = orjson.loads(self.request)
+        if "timestamp" not in d or d["timestamp"] is None:
+            d["timestamp"] = self.request_time
+
+        if "datasetId" not in d or d["datasetId"] is None:
+            _logger.error(f"Request missing dataset id {d}")
+            return None
+
+        if "embeddings" not in d or d["embeddings"] is None:
+            _logger.error(f"Request has no embeddings field {d}")
+            return None
+
+        if not isinstance(d["embeddings"], dict):
+            # TODO test recovering from errors like this. It seems to brick the container
+            _logger.error(
+                f'Expected a dictionary format for embeddings of the form {{"column_name": "embedding_2d_list"}}. Got {self.request}'
+            )
+            return None
+
+        return d
+
+
+def log_dict_to_data_frame(request: LogRequestDict) -> Tuple[pd.DataFrame, int]:
+    df = pd.DataFrame(request["multiple"]["data"], columns=request["multiple"]["columns"])
+    return df, len(df)
+
+
+def log_dict_to_embedding_matrix(request: LogEmbeddingRequestDict) -> Tuple[Dict[str, np.ndarray], int]:
+    row: Dict[str, np.ndarray] = {}
+    row_count = 0
+    for col, embeddings in request["embeddings"].items():
+        row[col] = np.array(embeddings)
+        row_count += len(embeddings)
+
+    return row, row_count
+
+
+def reduce_log_requests(acc: LogRequestDict, cur: LogRequestDict) -> LogRequestDict:
+    """
+    Reduce requests, assuming that each request has the same columns.
+    That assumption should be enforced before this is used by grouping by set of columns.
+    """
+    acc["multiple"]["data"].extend(cur["multiple"]["data"])
+    return acc
+
+
+def reduce_embeddings_request(acc: LogEmbeddingRequestDict, cur: LogEmbeddingRequestDict) -> LogEmbeddingRequestDict:
+    for col, embeddings in cur["embeddings"].items():
+        if col not in acc["embeddings"]:
+            acc["embeddings"][col] = []
+
+        acc["embeddings"][col].extend(embeddings)
+
+    return acc
+
+
+def determine_dataset_timestamp(
+    cadence: TimeGranularity, request: Union[LogRequestDict, LogEmbeddingRequestDict]
+) -> Optional[int]:
+    ts = request["timestamp"]
```

Comment:
it will break if "timestamp" is actually `None`. better with `request.get("timestamp")`

Reply:
from what I saw, you can define a `LogRequestDict` without a timestamp, since it's optional

Reply:
It doesn't break! It would only break if it were an untyped dict and the key didn't exist.  Added a test to confirm. A None value is still associated with the key.